### ISSUE TITLE
Reduce maintenance burden for go patch releases

### ISSUE THREAD

**Issue by jmhbnz** [etcd-io/etcd#15674]
### What would you like to be added?

Currently we specify which go patch version we build etcd with in 13 different files:

```bash
.github/workflows/tests-arm64.yaml:        go-version: "1.19.8"
.github/workflows/tests.yaml:        go-version: "1.19.8"
.github/workflows/govuln.yaml:          go-version: "1.19.8"
.github/workflows/grpcproxy.yaml:        go-version: "1.19.8"
.github/workflows/e2e.yaml:        go-version: "1.19.8"
.github/workflows/static-analysis.yaml:          go-version: "1.19.8"
.github/workflows/robustness-template.yaml:        go-version: '1.19.8'
.github/workflows/build.yaml:          go-version: "1.19.8"
.github/workflows/release.yaml:        go-version: "1.19.8"
.github/workflows/e2e-arm64.yaml:        go-version: "1.19.8"
.github/workflows/coverage.yaml:        go-version: "1.19.8"
.github/workflows/fuzzing.yaml:        go-version: "1.19.8"
.github/workflows/contrib.yaml:        go-version: "1.19.8"
````

Whenever a new go patch release comes out we have updated rapidly via pull request to ensure we are addressing CVE's, for example: #15329 #15427 #15649

I would like to propose that we manage this more autonomously by updating our workflows as follows:

```yaml
steps:
  - uses: actions/checkout@<hash>
  - uses: actions/setup-go@<hash>
    with:
      go-version-file: 'go.mod'
```

As outlined in the [setup-go action docs](https://github.com/actions/setup-go#getting-go-version-from-the-gomod-file) the `go-version-file` input accepts a path to a `go.mod` file or a `go.work` file that contains the version of Go to be used by a project. As the `go.mod` file contains only major and minor (e.g. 1.19) tags, the action will search for the latest available patch version sequentially in the runner's directory with the cached tools, in the [versions-manifest.json](https://github.com/actions/go-versions/blob/main/versions-manifest.json) file or at the go servers.

### Why is this needed?

This would reduce maintenance burden for etcd project, and also increase speed of adoption for new patch releases.

This also means we retain full decision making via pull request for the more important decision of when adopt new go major or minor releases and simplifies this process as we would just update `go.mod` files.

Additionally, in the unlikely event that a go patch release introduced build failures or other instability we can still do a pull request to pin to the previous patch release, as shown below, as if both the `go-version` and the `go-version-file` inputs are provided then the `go-version` input is used.

```diff
steps:
  - uses: actions/checkout@<hash>
  - uses: actions/setup-go@<hash>
    with:
      go-version-file: 'go.mod'
+     go-version: 1.19.8
```

So in summary this change would shift our approach from always requiring manual intervention when new go patch releases come out, to only requiring manual intervention when a new go patch comes out that causes problems.

**Comment by ahrtr** [comment]
Thanks @jmhbnz for raising this. Probably it makes more sense to follow [using-outputs-from-a-reusable-workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-outputs-from-a-reusable-workflow) to get go version (e.g. 1.19.8) defined in one place.

**Comment by jmhbnz** [comment]
> Thanks @jmhbnz for raising this. Probably it makes more sense to follow [using-outputs-from-a-reusable-workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-outputs-from-a-reusable-workflow) to get go version (e.g. 1.19.8) defined in one place.

Thanks @ahrtr, yes I did briefly look over that option.  I'm wondering if the two approaches combined could be even better? 

So yes we only want one file to update IF we want to pin a patch version (emphasis on the *if*). Then also because we have the re-usable workflow pointing at `go-version-file: 'go.mod'` we can also have the autonomy to always be using the latest patch release for whichever go release we are specifying in `go.mod`.

I think the main part of this proposal is the idea that we can avoid having to do 3x manual pull requests (`main` + 2x backports) whenever a new patch release of go comes out, and also react to that patch release automatically without any delay.

Will wait to see what maintainer consensus is before I raise anything. The centralizing of a pinned version via a re-usable workflow is still a good outcome if we think there is an issue with automatically adopting the latest patch release.

**Comment by ahrtr** [comment]
I suggest to manually & explictly set the go version (e.g. 1.19.8). If we link to go.mod file to always pickup the latest one, it might end up with unexpected release artifact which was built with the old go version.  

Let's work with an example, the workflows always pickup the latest go version thanks to linking to go.mod, then it may not detect the CVE described in [[security] Go 1.20.3 and Go 1.19.8 are released](https://groups.google.com/g/golang-announce/c/Xdv6JL9ENs8/m/OV40vnafAwAJ); so the release manager might not to update the go version on his release machine in time, eventually build & release etcd with 1.19.7.

**Comment by jmhbnz** [comment]
Thanks for the feedback on this.  So let's just limit this issue scope on seeing if it's possible to to centralise the current 13 values via re-usable workflows etc so there are less places to update.

**Comment by jmhbnz** [comment]
Centralizing our workflow `go-version` variable to one re-usable workflow could be a good issue for the upcoming etcd contribfest if there is someone interested in working on a ci improvements.

**Comment by judavi** [comment]
Hi! I'm interested! I'll be attending today to the contribfest.

**Comment by jmhbnz** [comment]
Completed via #15748
